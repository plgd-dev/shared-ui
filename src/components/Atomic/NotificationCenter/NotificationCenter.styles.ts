import { css } from '@emotion/react'
import { colors } from '../_utils/colors'
import { commonStyles, fontPrimary } from '../_utils/commonStyles'
import { ThemeType, get } from '../_theme'

export const floatingPanel = (theme: ThemeType) => css`
    z-index: 10;
    background: ${get(theme, `NotificationCenter.floatingPanel.background`)};
    border: 1px solid ${get(theme, `NotificationCenter.floatingPanel.borderColor`)};
    box-shadow: 0 30px 40px rgba(28, 52, 99, 0.1);
    border-radius: 8px;
    min-width: 400px;
    max-width: 600px;
`

export const header = css`
    padding: 16px;
    display: flex;
    height: 56px;
    box-sizing: border-box;
    align-items: center;
    justify-content: space-between;
    border-bottom: 1px solid #e6e9ed;
`

export const headline = (theme: ThemeType) => css`
    font-family: ${commonStyles.fontSecondary};
    font-style: normal;
    font-weight: 700;
    font-size: 16px;
    line-height: 140%;
    letter-spacing: -0.5px;
    color: ${get(theme, `NotificationCenter.Headline.color`)};
`

export const clearAll = (theme: ThemeType) => css`
    font-family: ${commonStyles.fontPrimary};
    font-style: normal;
    color: ${get(theme, `NotificationCenter.ClearAll.color`)};
    font-weight: 400;
    font-size: 12px;
    line-height: 18px;
    text-decoration: underline;

    &:hover {
        text-decoration: none !important;
        color: ${get(theme, `NotificationCenter.ClearAll.color`)};
    }
`

export const content = css`
    overflow: auto;
    max-height: 400px;
    padding: 0 16px;
`

export const noNotifications = css`
    padding: 16px;
    box-sizing: border-box;
    display: flex;
    align-items: center;
    justify-content: center;
    font-family: ${fontPrimary};
    font-style: normal;
    font-weight: 400;
    font-size: 14px;
    line-height: 22px;
    color: ${colors.neutral500};
`
