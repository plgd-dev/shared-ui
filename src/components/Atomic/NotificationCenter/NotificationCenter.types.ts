import { NotificationCenterItem } from 'react-toastify/addons/use-notification-center'

export type NotificationCenterItemType = Partial<NotificationCenterItem> & {
    text: string
}

export type Props = {
    i18n: {
        notifications: string
        noNotifications: string
        markAllAsRead: string
    }
}

export const defaultProps: Partial<Props> = {
    i18n: {
        notifications: 'Notifications',
        noNotifications: 'No notifications',
        markAllAsRead: 'mark all as read',
    },
}
