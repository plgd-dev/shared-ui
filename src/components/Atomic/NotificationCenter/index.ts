export { default } from './NotificationCenter'
export * from './NotificationCenter'
