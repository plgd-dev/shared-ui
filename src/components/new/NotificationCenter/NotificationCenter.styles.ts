import { css } from '@emotion/react'
import { colors } from '../_utils/colors'
import { commonStyles } from '../_utils/commonStyles'

export const floatingPanel = css`
    z-index: 10;
    background: #fff;
    border: 1px solid #e6e9ed;
    box-shadow: 0 30px 40px rgba(28, 52, 99, 0.1);
    border-radius: 8px;
    min-width: 400px;
`

export const header = css`
    padding: 16px;
    display: flex;
    align-items: center;
    justify-content: space-between;
`

export const headline = css`
    font-family: ${commonStyles.fontSecondary};
    font-style: normal;
    font-weight: 700;
    font-size: 16px;
    line-height: 140%;
    letter-spacing: -0.5px;
`

export const clearAll = css`
    font-family: ${commonStyles.fontPrimary};
    font-style: normal;
    color: ${colors.primary};
    font-weight: 400;
    font-size: 12px;
    line-height: 18px;
    text-decoration: underline;

    &:hover {
        text-decoration: none;
    }
`

export const content = css``

export const item = css``
