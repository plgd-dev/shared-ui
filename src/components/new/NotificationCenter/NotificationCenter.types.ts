import { NotificationCenterItem } from "react-toastify/addons/use-notification-center";

export type NotificationCenterItemType = Partial<NotificationCenterItem> & {
  text: string
}

export type Props = {}
