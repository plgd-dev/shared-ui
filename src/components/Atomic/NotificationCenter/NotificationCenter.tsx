import { FC, useCallback, useEffect, useRef, useState } from 'react'
import { offset, shift, useFloating } from '@floating-ui/react'
import { motion, AnimatePresence } from 'framer-motion'
import { FloatingPortal } from '@floating-ui/react-dom-interactions'
import { useNotificationCenter } from 'react-toastify/addons/use-notification-center'
import isFunction from 'lodash/isFunction'

import { Props, defaultProps } from './NotificationCenter.types'
import Bell from './components/Bell'
import * as styles from './NotificationCenter.styles'
import { hasEventBlocker } from '../_utils/envets'
import InnerToast from './components/InnerToast/InnerToast'
import Scrollbars from '../Scrollbars'

const NotificationCenter: FC<Props> = (props) => {
    const { defaultNotification, i18n, onNotification, readAllNotifications } = { ...defaultProps, ...props }
    const [open, setOpen] = useState(false)
    const ref = useRef(null)
    const { x, y, refs, strategy } = useFloating({
        placement: 'bottom-end',
        strategy: 'fixed',
        middleware: [shift(), offset(4)],
    })

    const notificationsCount = useRef<number>(defaultNotification?.length ?? 0)
    // const notificationsCount = useRef(1)
    const { notifications, markAllAsRead, unreadCount } = useNotificationCenter({
        data: defaultNotification ?? [],
        sort: (l, r) => r.createdAt - l.createdAt,
    })

    const readAll = useCallback(() => {
        isFunction(readAllNotifications) && readAllNotifications()
        markAllAsRead()
    }, [markAllAsRead, readAllNotifications])

    useEffect(() => {
        function handleClickOutside(event: any) {
            // @ts-ignore
            if (ref?.current && !ref?.current?.contains(event.target) && !hasEventBlocker(event, 'data-block-outside-click')) {
                setOpen(false)
            }
        }
        if (open) {
            document.addEventListener('mousedown', handleClickOutside)
            return () => {
                document.removeEventListener('mousedown', handleClickOutside)
            }
        }
    }, [ref, open])

    useEffect(() => {
        if (notifications && notifications.length !== notificationsCount.current) {
            notificationsCount.current = notifications.length
            isFunction(onNotification) && notifications && onNotification(notifications)
        }
    }, [notifications, onNotification])

    return (
        <div>
            <Bell hasUnRead={unreadCount > 0} innerRef={refs.setReference} notificationsCount={notifications.length} onClick={() => setOpen(!open)} />
            <FloatingPortal>
                <AnimatePresence>
                    {open && (
                        <motion.div
                            animate={{ opacity: 1, scale: 1, y: 0 }}
                            css={styles.floatingPanel}
                            exit={{ opacity: 0, y: 0 }}
                            initial={{ opacity: 0, scale: 0.85, y: -100 }}
                            ref={refs.setFloating}
                            style={{
                                position: strategy,
                                top: y ?? 0,
                                left: x ?? 0,
                                width: 'max-content',
                            }}
                            transition={
                                // When in "grouped phase", make the transition faster
                                // The open delay becomes 1ms during this phase.
                                { type: 'spring', damping: 20, stiffness: 300 }
                            }
                        >
                            <div ref={ref}>
                                <div css={styles.header}>
                                    <div css={styles.headline}>{i18n.notifications}</div>
                                    {notifications.length > 0 && (
                                        <a
                                            css={styles.clearAll}
                                            href='#'
                                            onClick={(e) => {
                                                e.preventDefault()
                                                readAll()
                                            }}
                                        >
                                            {i18n.markAllAsRead}
                                        </a>
                                    )}
                                </div>
                                <AnimatePresence>
                                    <Scrollbars autoHeight autoHeightMax={400} autoHeightMin={50}>
                                        <motion.section
                                            animate={open ? 'open' : 'closed'}
                                            css={styles.content}
                                            variants={{
                                                open: {
                                                    transition: { staggerChildren: 0.07, delayChildren: 0.2 },
                                                },
                                                closed: {
                                                    transition: { staggerChildren: 0.05, staggerDirection: -1 },
                                                },
                                            }}
                                        >
                                            <AnimatePresence>
                                                {notifications.length === 0 && <div css={styles.noNotifications}>{i18n.noNotifications}</div>}
                                                {notifications.length > 0 &&
                                                    notifications.map((notification: any, i) => (
                                                        <InnerToast borderTop={i !== 0} key={i} notification={notification} />
                                                    ))}
                                            </AnimatePresence>
                                        </motion.section>
                                    </Scrollbars>
                                </AnimatePresence>
                            </div>
                        </motion.div>
                    )}
                </AnimatePresence>
            </FloatingPortal>
        </div>
    )
}

NotificationCenter.displayName = 'NotificationCenter'
NotificationCenter.defaultProps = defaultProps

export default NotificationCenter
