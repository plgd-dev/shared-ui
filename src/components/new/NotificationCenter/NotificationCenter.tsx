import { FC, useState } from 'react'
import { NotificationCenterItemType, Props } from "./NotificationCenter.types";
import Bell from './components/Bell'
import { offset, shift, useFloating } from '@floating-ui/react'
import * as styles from './NotificationCenter.styles'
import { FloatingPortal } from '@floating-ui/react-dom-interactions'
import { motion, AnimatePresence } from 'framer-motion'
import { NotificationCenterItem, useNotificationCenter } from 'react-toastify/addons/use-notification-center'

const NotificationCenter: FC<Props> = () => {
    const [open, setOpen] = useState(false)
    const { x, y, reference, floating, strategy, context } = useFloating({
        placement: 'bottom-end',
        strategy: 'fixed',
        middleware: [shift(), offset(4)],
    })
    const { notifications, clear, markAllAsRead, markAsRead, remove, unreadCount } = useNotificationCenter()
    console.log(notifications)
    return (
        <div>
            <Bell onClick={() => setOpen(!open)} hasUnRead={true} innerRef={reference} />
            <FloatingPortal>
                <AnimatePresence>
                    {open && (
                        <motion.div
                            animate={{ opacity: 1, scale: 1, y: 0 }}
                            className='tooltip-bubble'
                            exit={{ opacity: 0, y: 0 }}
                            initial={{ opacity: 0, scale: 0.85, y: -100 }}
                            css={styles.floatingPanel}
                            ref={floating}
                            style={{
                                position: strategy,
                                top: y ?? 0,
                                left: x ?? 0,
                                width: 'max-content',
                            }}
                            transition={
                                // When in "grouped phase", make the transition faster
                                // The open delay becomes 1ms during this phase.
                                { type: 'spring', damping: 20, stiffness: 300 }
                            }
                        >
                            <div css={styles.header}>
                                <div css={styles.headline}>Notifications</div>
                                <a href='#' css={styles.clearAll}>
                                    mark all as read
                                </a>
                            </div>
                            <AnimatePresence>
                                <motion.section
                                    css={styles.content}
                                    variants={{
                                        open: {
                                            transition: { staggerChildren: 0.07, delayChildren: 0.2 },
                                        },
                                        closed: {
                                            transition: { staggerChildren: 0.05, staggerDirection: -1 },
                                        },
                                    }}
                                    animate={open ? 'open' : 'closed'}
                                >
                                    <AnimatePresence>
                                        {notifications.map((notification: any) => (
                                            <motion.div
                                                key={notification.id}
                                                layout
                                                initial={{ scale: 0.4, opacity: 0, y: 50 }}
                                                exit={{
                                                    scale: 0,
                                                    opacity: 0,
                                                    transition: { duration: 0.2 },
                                                }}
                                                animate={{ scale: 1, opacity: 1, y: 0 }}
                                            >
                                                <motion.article
                                                    variants={{
                                                        open: {
                                                            y: 0,
                                                            opacity: 1,
                                                            transition: {
                                                                y: { stiffness: 1000, velocity: -100 },
                                                            },
                                                        },
                                                        closed: {
                                                            y: 50,
                                                            opacity: 0,
                                                            transition: {
                                                                y: { stiffness: 1000 },
                                                            },
                                                        },
                                                    }}
                                                    css={styles.item}
                                                >
                                                    {notification.data?.text}
                                                </motion.article>
                                            </motion.div>
                                        ))}
                                    </AnimatePresence>
                                </motion.section>
                            </AnimatePresence>
                        </motion.div>
                    )}
                </AnimatePresence>
            </FloatingPortal>
        </div>
    )
}

NotificationCenter.displayName = 'NotificationCenter'

export default NotificationCenter
