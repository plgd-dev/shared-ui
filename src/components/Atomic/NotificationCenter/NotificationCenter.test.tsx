import { render } from '@testing-library/react'
import NotificationCenter from './NotificationCenter'

describe('<NotificationCenter>', () => {
    it('should render without crashing', () => {
        const { container, asFragment } = render(
            <NotificationCenter i18n={{ notifications: 'notifications', noNotifications: 'noNotifications', markAllAsRead: 'markAllAsRead' }} />
        )
        expect(container).toBeTruthy()
        expect(asFragment()).toMatchSnapshot()
    })
})
