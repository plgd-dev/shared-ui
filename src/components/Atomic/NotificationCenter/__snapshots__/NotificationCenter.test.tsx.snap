// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NotificationCenter> should render without crashing 1`] = `
<DocumentFragment>
  .emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-transition: all 0.3s;
  transition: all 0.3s;
  cursor: pointer;
}

.emotion-1 {
  width: 44px;
  height: 44px;
  border-radius: 50%;
  position: relative;
  -webkit-transition: all 0.3s;
  transition: all 0.3s;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
}

.emotion-1 .shake {
  -webkit-animation: bellshake 0.5s cubic-bezier(0.36, 0.07, 0.19, 0.97) both;
  animation: bellshake 0.5s cubic-bezier(0.36, 0.07, 0.19, 0.97) both;
  -webkit-backface-visibility: hidden;
  backface-visibility: hidden;
  transform-origin: top right;
}

<div>
    <span
      class="emotion-0"
      data-block-outside-click="true"
    >
      <a
        class="emotion-1"
        data-block-outside-click="true"
        href="#"
      >
        <svg
          fill="none"
          height="24"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M12 2c-.711 0-1.286.559-1.286 1.25V4c-2.933.578-5.142 3.102-5.142 6.125v.734a7.4 7.4 0 0 1-1.949 4.985l-.297.324a1.23 1.23 0 0 0-.213 1.344c.205.449.667.738 1.173.738h15.429c.506 0 .964-.29 1.173-.738a1.22 1.22 0 0 0-.213-1.344l-.298-.324a7.4 7.4 0 0 1-1.948-4.985v-.734c0-3.023-2.21-5.547-5.143-6.125v-.75C13.286 2.559 12.71 2 12 2m1.82 19.27a2.47 2.47 0 0 0 .752-1.77H9.429c0 .664.269 1.3.751 1.77A2.6 2.6 0 0 0 12 22a2.6 2.6 0 0 0 1.82-.73"
            fill="currentcolor"
          />
        </svg>
      </a>
    </span>
  </div>
</DocumentFragment>
`;
